src/back/msl/writer.rs

```diff
@@ -1192,7 +1195,67 @@ impl<W: Write> Writer<W> {
                     Mf::Log2 => "log2",
                     Mf::Pow => "pow",
                     // geometry
-                    Mf::Dot => "dot",
+                    Mf::Dot => {
+                        use crate::TypeInner;
+                        let inner = context.resolve_type(arg);
+                        match *inner {
+                            TypeInner::Vector { kind, size, .. } => {
+                                use crate::ScalarKind;
+                                match kind {
+                                    ScalarKind::Float => "dot",
+                                    ScalarKind::Sint | ScalarKind::Uint => {
+                                        // No intrinsic function for integer dot product in MSL,
+                                        // transform the function call into an arithmetic expression.
+
+                                        // arg1 should be safe to unwrap due to previous validation
+                                        let arg1 = arg1.unwrap();
+
+                                        write!(self.out, "(")?;
+
+                                        self.put_expression(arg, context, false)?;
```

Comment:
since the backend guarantees `arg` and `arg1` to be baked here, we can and should write this directly:
```rust
        let arg0_name = self.named_expressions[&arg];
        let arg1_name = self.named_expressions[&arg1];
        write!(self.out, "+ {}.{} * {}.{}", arg0_name, component, arg1_name, component)?;
```